### PR TITLE
sql: do not hide columns in introspection tables

### LINKED ISSUE #28548: sql: Hide implicit rowid index from introspection interfaces
The behavior of the various introspection interfaces (I'm mainly concerned with information_schema, but show columns/indexes behaves similarly. haven't checked pg_catalog) is inconsistent with respect to the hidden rowid column and associated primary key. The column is hidden from information_schema.columns, but the index appears in information_schema.statistics. This confuses some clients, including SQLAlchemy 1.2's tests, which fail because there is an index which refers to a column that cannot be found. If we hide the column from information_schema.columns, we should also hide the index from information_schema.indexes.

```
root@127.0.0.1:58397/defaultdb> select * from information_schema.columns where table_name='nopk';
  table_catalog | table_schema | table_name | column_name | ordinal_position | column_default | is_nullable | data_type | character_maximum_length | character_octet_length | numeric_precision | numeric_precision_radix | numeric_scale | datetime_precision | character_set_catalog | character_set_schema | character_set_name | generation_expression  
+---------------+--------------+------------+-------------+------------------+----------------+-------------+-----------+--------------------------+------------------------+-------------------+-------------------------+---------------+--------------------+-----------------------+----------------------+--------------------+-----------------------+
  defaultdb     | public       | nopk       | a           |                1 | NULL           | YES         | INT       |                     NULL |                   NULL |                64 |                       2 |             0 |               NULL | NULL                  | NULL                 | NULL               |                        
(1 row)

Time: 3.694411ms

root@127.0.0.1:58397/defaultdb> select * from information_schema.statistics where table_name='nopk';;
  table_catalog | table_schema | table_name | non_unique | index_schema | index_name | seq_in_index | column_name | COLLATION | cardinality | direction | storing | implicit  
+---------------+--------------+------------+------------+--------------+------------+--------------+-------------+-----------+-------------+-----------+---------+----------+
  defaultdb     | public       | nopk       | NO         | public       | primary    |            1 | rowid       | NULL      |        NULL | ASC       | NO      | NO        
(1 row)

Time: 1.81012ms
```

### PR DESCRIPTION
**Pull request by knz**
Fixes #28548.

Requested/suggested by @bdarnell

Clients were confused by the fact some items in introspection referred
to columns that were otherwise not listed because they were marked as
hidden in the table descriptors. For example, an index would
refer to a column (rowid) not listed in the columns for that table.

This has been found (by Ben) to be a problem for SQLAlchemy and (by
me) in dBeaver.

In contrast, PostgreSQL which also has hidden columns always shows all
columns in introspection, including hidden columns, in both
information_schema and pg_catalog.

This patch corrects the issue by doing the same as PostgreSQL.  The
output of SHOW COLUMNS is extended to indicate which columns are
hidden.

Release note (bug fix): Hidden columns are now listed in
`information_schema` and `pg_catalog` tables, for better compatibility
with PostgreSQL.

Release note (sql change): The output of `SHOW COLUMNS` is extended to
indicate which columns are hidden.

### CODE REVIEW

**Review by jordanlewis — COMMENTED** [review]
Yikes, this looks like it breaks one of the example orms checks...

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28750#-:-LK7OCWw3wbnA_AdKvk6:b-cpi0ju)* status: :shipit: complete! 0 of 0 LGTMs obtained

File: pkg/sql/information_schema.go [hunk 1]
```diff
@@ -254,7 +254,8 @@ CREATE TABLE information_schema.columns (
 	CHARACTER_SET_CATALOG    STRING,
 	CHARACTER_SET_SCHEMA     STRING,
 	CHARACTER_SET_NAME       STRING,
-	GENERATION_EXPRESSION    STRING
+	GENERATION_EXPRESSION    STRING,
+	IS_HIDDEN                STRING NOT NULL -- CockroachDB extension
```
Comment (jordanlewis) [inline pkg/sql/information_schema.go]: I have a medium-strength objection to this.  It's trouble to add things to information_schema unless we absolutely have to. I would prefer to leave information_schema.columns the same, and let the hidden-ness be ambiguous, just like it is in Postgres.

**Review by jordanlewis — APPROVED** [review]
LGTM mod comment, thanks!

**Comment by knz** [comment]
The failing test was because example-orms needed a boost (you approved that PR). The tests should pass now.

I do not understand your concern about information_schema. Everyone seems to place their extensions in here:

- mysql `generation_expression` etc: https://dev.mysql.com/doc/refman/5.7/en/columns-table.html
- mssql `domain_catalog` etc: https://www.mssqltips.com/sqlservertutorial/183/informationschemacolumns/
- pg `udt_catalog` etc: https://www.postgresql.org/docs/10/static/infoschema-columns.html

What's the trouble exactly?

**Comment by jordanlewis** [comment]
I didn't realize you actually needed the column. In that case, I think I'm fine with it. LGTM!

**Comment by knz** [comment]
thank you! :sparkling_heart: 

bors r+

**Pull request merged**